<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>继发和异步</title>
    <link rel="stylesheet" href="../css/main.css" />
    <link
      href="https://cdn.bootcdn.net/ajax/libs/font-awesome/5.14.0/css/all.css"
      rel="stylesheet"
    />
    <style>
      * {
        transition: all 0.3s ease;
      }
      body {
        /* DES 就算超出 main 也不会很突兀 ，因为 Body 有底色 */
        background-color: #126993;
        font-family: Modak, source-han-serif-sc, serif, 'Nunito', 'Lato', consolas;
      }
      .main_container {
        min-height: 100vh;
        color: white;
        font-size: 1.6rem;
        user-select: none;
      }
      .waterfall_container {
        min-height: 50%;
        min-width: 50vw;
        max-width: 500px;
        padding: 0 5px;
        box-sizing: border-box;
        background-color: #426a5b;
      }
      .img-list {
        list-style: none;
      }
      .img-list .item {
        margin: 10px 0;
        padding: 3px;
        box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
        border: 1px solid white;
        background-color: white;
      }
      .img-list img {
        height: 100px;
        filter: brightness(0.8);
        vertical-align: middle;
        object-fit: contain;
      }
      .img-list img:hover {
        filter: brightness(1);
      }
    </style>
  </head>
  <body>
    <main class="main_container m-flex flex-jc-center flex-an-center flex-dir-col">
      <h2>继发和异步</h2>
      <h4>函数reqOne()，完成之后才能调用 reqTwo()</h4>
      <h4>实现方式，循环</h4>
    </main>
  </body>
  <script>
    /*
      查看控制器可以得出，抛出异常后，reqTwo 没有继续发起请求，说名是继发
    */
    let urls = [
      'https://tse1-mm.cn.bing.net/th/id/OIP.S19cENn0MaOaS_DwBqBrbAHaE6?pid=Api&rs=1',
      'https://tse1-mm.cn.bing.net/th/id/OIP.FM9a5hrdTWFVVpMGyq_LUgHaE8?pid=Api&rs=1'
    ]
    let res = ''
    async function reqOne() {
      // throw new Error('one error')
      res = await fetch(urls[0])
      console.log('reqOne 完成!')
      return
    }
    async function reqTwo() {
      // throw new Error('two error')
      await fetch(urls[1] + res)
      console.log('reqTwo 完成!', res)
      return
    }

    async function loopReq(...fns) {
      console.log('当前需要继发的函数有:', fns)
      let len = fns.length
      for (fn of fns) {
        await fn()
      }
    }

    async function reqThree(arg) {
      res = await fetch(urls[0])
      console.log('reqThree 完成!，我的顺序不可信!', arg)
      return arg + ',reqThree return'
    }
    async function reqFour(arg) {
      await fetch(urls[1])
      console.log('reqFour 完成!，我的顺序不可信!', arg)
      return (await arg) + ',reqFour return'
    }
    async function reqFive(arg) {
      await fetch(urls[1])
      console.log('reqFive 完成!，我的顺序不可信!', arg)
      return (await arg) + ',reqFive return'
    }

    function compose(...fns) {
      return async function (x) {
        return await fns.reduce(async (arg, fn) => {
          await arg
          return fn(arg)
        }, x)
      }
    }
    let test = compose(reqThree, reqFour, reqFive)
    test('run').then(async (res) => {
      console.log('compose done:', res)
    })
    // loopReq(reqOne, reqTwo)
  </script>
</html>
